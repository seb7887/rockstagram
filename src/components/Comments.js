import React from 'react';
import { Comment, CommentGrid } from './styles/CommentStyles';

class Comments extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      author: '',
      comment: ''
    };
  }

  renderComment(comment, i) {
    return (
      <Comment key={i} data-testid='comment'>
        <p>
          <strong>{comment.user}</strong>
          {comment.text}
          <button
            data-testid='remove-comment'
            className='remove-comment'
            onClick={this.props.removeComment.bind(
              null,
              this.props.match.params.postId,
              i,
            )}
          >
            &times;
          </button>
        </p>
      </Comment>
    );
  }

  handleChange = e => {
    e.preventDefault();
    this.setState({ [e.target.name]: e.target.value });
  }

  handleSubmit = e => {
    e.preventDefault();
    this.props.addComment(
      this.props.match.params.postId,
      this.state.author,
      this.state.comment,
    );
    this.clearForm();
    this.setState({
      author: '',
      comment: ''
    });
  };

  clearForm() {
    document.getElementById('form').reset();
  }

  render() {
    const comments = this.props.comments[this.props.match.params.postId] || [];
    return (
      <CommentGrid>
        {comments.map(this.renderComment.bind(this))}

        <form
          onSubmit={this.handleSubmit}
          id='form'
          className='comment-form'
          data-testid='form'
        >
          <input
            type='text'
            name='author'
            placeholder='Author'
            onChange={this.handleChange}
            data-testid='author'
          />
          <input
            type='text'
            name='comment'
            placeholder='Comment here...'
            onChange={this.handleChange}
            data-testid='text'
          />
          <input type='submit' hidden />
        </form>
      </CommentGrid>
    );
  }
}

export default Comments;
