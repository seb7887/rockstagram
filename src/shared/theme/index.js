const theme = {
  colors: {
    black: '#0b0b0b',
    offwhite: '#ffffff',
    white: 'white',
    lightgrey: '#fafafa',
    red: 'red',
    blue: '#5995ef',
    grey: '#e6e6e6',
  },
  fonts: {
    // eslint-disable-next-line quotes
    header: `billabong, 'billabongregular'`,
    p: 'Segoe UI',
  },
};

export default theme;
