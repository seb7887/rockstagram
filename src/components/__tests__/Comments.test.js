/* eslint-env jest */
import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render, fireEvent } from 'react-testing-library';

import Comments from '../Comments';
import theme from '../../shared/theme';

let comments = {
  test: [
    {
      text: 'This is a test',
      user: 'kingmob',
    },
  ],
};

const removeComment = jest.fn();
const addComment = jest.fn();

let mockedProps = {
  match: {
    params: {
      postId: 'test',
    },
  },
  comments,
  removeComment,
  addComment,
};

describe('<Comments/>', () => {
  it('renders and matches snapshot', () => {
    const { asFragment } = render(
      <ThemeProvider theme={theme}>
        <Comments {...mockedProps} />
      </ThemeProvider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('displays comments correctly', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Comments {...mockedProps} />
      </ThemeProvider>,
    );

    expect(getByTestId('comment').textContent).toContain('This is a test');
  });

  it('adds a new comment', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Comments {...mockedProps} />
      </ThemeProvider>,
    );

    fireEvent.change(getByTestId('author'), { target: { value: 'test' } });
    fireEvent.change(getByTestId('text'), { target: { value: 'hey!' } });

    fireEvent.submit(getByTestId('form'));

    expect(addComment).toHaveBeenCalledTimes(1);
  });

  it('removes a comment', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Comments {...mockedProps} />
      </ThemeProvider>,
    );

    fireEvent.click(getByTestId('remove-comment'));

    expect(removeComment).toHaveBeenCalledTimes(1);
  });
});
