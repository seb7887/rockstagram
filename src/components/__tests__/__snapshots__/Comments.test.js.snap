// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comments/> renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="CommentStyles__CommentGrid-qqfm26-1 gNQaPS"
  >
    <div
      class="CommentStyles__Comment-qqfm26-0 bcqYLs"
      data-testid="comment"
    >
      <p>
        <strong>
          kingmob
        </strong>
        This is a test
        <button
          class="remove-comment"
          data-testid="remove-comment"
        >
          Ã—
        </button>
      </p>
    </div>
    <form
      class="comment-form"
      data-testid="form"
      id="form"
    >
      <input
        data-testid="author"
        name="author"
        placeholder="Author"
        type="text"
      />
      <input
        data-testid="text"
        name="comment"
        placeholder="Comment here..."
        type="text"
      />
      <input
        hidden=""
        type="submit"
      />
    </form>
  </div>
</DocumentFragment>
`;
